compositor_render/src/transformations/builtin/transform_to_resolution/fill.wgsl

```diff
@@ -0,0 +1,78 @@
+struct VertexInput {
+    @location(0) position: vec3<f32>,
+    @location(1) tex_coords: vec2<f32>,
+}
+
+struct VertexOutput {
+    @builtin(position) position: vec4<f32>,
+    @location(0) tex_coords: vec2<f32>,
+}
+
+struct CommonShaderParameters {
+    time: f32,
+    textures_count: u32,
+    output_resolution: vec2<u32>,
+}
+
+var<push_constant> common_params: CommonShaderParameters;
+
+@group(0) @binding(0) var textures: binding_array<texture_2d<f32>, 16>;
+@group(2) @binding(0) var sampler_: sampler;
+
+fn get_scale_matrix(x_scale: f32, y_scale: f32) -> mat4x4<f32> {
+    let col1: vec4<f32> = vec4<f32>(x_scale, 0.0, 0.0, 0.0); 
+    let col2: vec4<f32> = vec4<f32>(0.0, y_scale, 0.0, 0.0); 
+    let col3: vec4<f32> = vec4<f32>(0.0, 0.0, 1.0, 0.0); 
+    let col4: vec4<f32> = vec4<f32>(0.0, 0.0, 0.0, 1.0); 
+    return mat4x4<f32>(col1, col2, col3, col4);
+}
+
+@vertex
+fn vs_main(input: VertexInput) -> VertexOutput {
+    var output: VertexOutput;
+    let input_texture_size: vec2<u32> = textureDimensions(textures[0]);
+    let input_texture_width: f32 = f32(input_texture_size.x);
+    let input_texture_height: f32 = f32(input_texture_size.y);
+    let input_ratio: f32 = input_texture_width / input_texture_height;
+
+    let output_texture_width: f32 = f32(common_params.output_resolution.x);
+    let output_texture_height: f32 = f32(common_params.output_resolution.y);
+    let output_ratio: f32 = output_texture_width / output_texture_height;
+
+    var x_scale: f32 = 1.0;
+    var y_scale: f32 = 1.0;
+
+    // This transformation preserves the input texture ratio.
+    //
+    // If the input ratio is larger than the output ratio, the texture is scaled,
+    // such that input height = output height. Then:
+    // scale_factor_pixels = output_height / input_height
+    // Using clip space coords ([-1, 1] range in both axis):
+    // scale_factor_x_clip_space = scale_factor_pixels * input_width / output_width
+    // scale_factor_x_clip_space = (output_height * input_width) / (output_width * input_height)
+    // scale_factor_x_clip_space = input_ratio / output_ratio
+    // scale_factor_y_clip_space = 1.0 (input y coords are already fitted)
+    //
+    // If the output ratio is larger, then the texture is scaled up,
+    // such that input_width = output_width.
+    // Analogously:
+    // scale_factor_x_clip_space = 1.0 (input x coords are already fitted)
+    // scale_factor_y_clip_space = output_ratio / input_ratio
+    if input_ratio >= output_ratio {
+        x_scale = input_ratio / output_ratio;
+    } else {
+        y_scale = output_ratio / input_ratio;
+    }
```

Comment:
I don't think this works, we should discuss this in person

Reply:
how can this preserve the aspect ratio if you multiply one coordinate by some number != 1 and the other by 1?

Reply:
Since those are output texture coords :)